### PR TITLE
[30%] HSEARCH-1872 Create a proper API for sorting definitions

### PR DESCRIPTION
**Pull request by emmanuelbernard**
This PR shows the interface needed to implement the DSL.
One can already play with it and compile the test code, not run it.

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortContext.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	 * Order elements by a specific field
+	 */
+	SortFieldContext byField(String field);
+
+	/**
+	 * Sort by distance from an already built spatial query.
+	 * For distance sort without a query, use {@code byField("location").fromCoordinates(...)}
+	 * or {@code byField("location").fromLatitude(...)}
+	 */
+	SortDistanceContext byDistanceFromSpatialQuery(Query query);
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortContext.java]: here I'm wondering if we could avoid references to the Lucene type 'query'.

Assuming in a next step we'll be able to control the Sort DSL directly on the Query Builder (the existing Query DSL), then such a thing would go away right?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.query.dsl.sort;
+
+/**
+ * @author Emmanuel Bernard emmanuel@hibernate.org
+ */
+public interface SortNativeContext extends SortAdditionalSortFieldContext, SortTermination {
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: I guess you could persuade me to introduce a `SortNativeContext`, but I wasn't expecting us to provide such a thing.

Shouldn't we encourage people needing "native access" to avoid the DSL?

Do you think we should actually deprecate the old "setSort" and have all users - including the ones with specific native needs - move to this DSL?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortFieldContext.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	 * Latitude in degree
+	 */
+	SortLatLongContext fromLatitude(double latitude);
+
+	// native comparator
+
+	/**
+	 * Compare field with specific comparator implementation.
+	 */
+	SortComparatorContext withComparator(Comparator<?> comparator);
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortFieldContext.java]: This is one of those things which maybe we shouldn't expose? Hard to abstract across different search engines.

Happy to leave it if you have a solid use case.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @author Emmanuel Bernard emmanuel@hibernate.org
+ */
+public interface SortDistanceContext extends SortOrder<SortDistanceContext>, SortAdditionalSortFieldContext, SortTermination {
+
+	/**
+	 * Hint for the backend to return data in the appropriate unit.
+	 */
+	SortDistanceContext in(Unit unit);
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java]: I'm not sure what "to return data" is referring to. Shouldn't this be an option for projections?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	 * Sort by distance from an already built spatial query.
+	 * For distance sort without a query, use {@code byField("location").fromCoordinates(...)}
+	 * or {@code byField("location").fromLatitude(...)}
+	 */
+	SortDistanceContext andByDistanceFromSpatialQuery(Query query);
+
+	/**
+	 * Order element using the native backend API for Lucene.
+	 */
+	SortNativeContext andByNative(SortField sortField);
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: I'd remove the following two which return `SortNativeContext` for now. Happy to re-add such things later?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	 * This is the fastest sorting.
+	 */
+	SortOrderTermination andByIndexOrder();
+
+	/**
+	 * Sort by distance from an already built spatial query.
+	 * For distance sort without a query, use {@code byField("location").fromCoordinates(...)}
+	 * or {@code byField("location").fromLatitude(...)}
+	 */
+	SortDistanceContext andByDistanceFromSpatialQuery(Query query);
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: I'm confused by this one. What does it mean to sort the current Query A "by distance from an already built spatial query" ?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	/**
+	 * Order elements by a specific field
+	 */
+	SortFieldContext andByField(String field);
+
+	/**
+	 * Order elements by their relevance score in decreasing order by default.
+	 */
+	//TODO Is byScore terminal or can it be followed by other sorts?
+	SortOrderTermination andByScore();
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: I'd remove this option for now as mixing "by score" with an additional sorting option is quite exotic.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+	/**
+	 * Order elements by their internal index order.
+	 * This is the fastest sorting.
+	 */
+	SortOrderTermination andByIndexOrder();
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: Similarly to _andByScore_, I'd remove  this option too as it makes even less sense to use in practice to mix another sorting with the "by index" sort.

I think "by index" is used exclusively for fastest speed when sorting doesn't matter at all (in which case one wouldn't combine it with another sort), or if the index was specifically constructed to be laid out in a specific sort order but we don't give such control.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+	/**
+	 * Hint for the backend to use a specific distance algorithm
+	 */
+	SortDistanceContext withComputeMethod(DistanceMethod distanceMethod);
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java]: We currently support two modes {{SpatialMode.RANGE}} and {{SpatialMode.HASH}}, but these control indexing format, so a query needs to match the chosen type.

For sorting, we only have one strategy which is based on `org.hibernate.search.spatial.impl.DistanceComparator.DistanceComparator(Point, int, String)`.. how does this fit in, are you planning to provide more sorting implementation alternatives?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ */
+package org.hibernate.search.query.dsl.sort;
+
+/**
+ * Mode when comparing multiple values.
+ *
+ * @author Emmanuel Bernard emmanuel@hibernate.org
+ */
+public enum MultiValuedMode {
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java]: Nice, hadn't thought about multi-values. I'm afraid you'll have to pick a default, so that it can work even if the user didn't  invoke the method `org.hibernate.search.query.dsl.sort.SortFieldContext.treatMultiValuedAs(MultiValuedMode)` ?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+	/**
+	 * Pick the lowest value.
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java]: I understand what you mean here, but this should probably rephrase to clarify how it would work on Strings.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+	MIN,
+
+	/**
+	 * Pick the highest value.
+	 */
+	MAX,
+
+	/**
+	 * Use the sum of all values as sort value.
+	 * Only applicable for number based array fields.
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java]: "Only applicable for number based array fields"

I guess the focus should be on *number* but not strictly only an *array* ? Could work also on numeric values which have been indexed as strings via some string bridge?

Comment applies on all three of SUM, AVG, MEDIAN.

BTW as commented on JIRA, in this case it seems like you're also adding a new feature (on top of a new API).. feel free to split this up into a follow up task? I think currently our docs are pretty clear that you can't sort on multi-valued fields.

File: orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+		cfg.put( Environment.MODEL_MAPPING, MappingFactory.class.getName() );
+	}
+
+	public static class MappingFactory {
+		@Factory
+		public SearchMapping build() {
+			SearchMapping mapping = new SearchMapping();
+			mapping
+					.analyzerDef( "stemmer", StandardTokenizerFactory.class )
+					.filter( StandardFilterFactory.class )
```
Comment (Sanne) [inline orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java]: you'd normally indent such things nicely.. did you give up to auto-formatters? ;-)

File: orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+				Animal.class,
+				Day.class,
+				CoffeeBrand.class,
+				Coffee.class
+		};
+	}
+
+	@Override
+	public void configure(Map<String,Object> cfg) {
+		cfg.put( Environment.MODEL_MAPPING, MappingFactory.class.getName() );
```
Comment (Sanne) [inline orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java]: unrelated, but FYI you can put the `SearchMapping` instance directly in the cfg, so to not have to fiddle with classname, @Factory annotations, etc..

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortContext.java (hunk 1 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortContext.java]: When we generalise the DSL, this would naturally be using the right generic type. I did not see a point in working on that now.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java (hunk 2 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: In the 5.x series setSort cannot be deprecated. The DSL output is what is passed to setSort. In 6 all bets are off but I am not sure how to do it. 

I personally like how one can mix normal and native declarations. But we can remove it if people don't agree with me.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortFieldContext.java (hunk 3 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortFieldContext.java]: It might be harder than I hoped so we will likely drop it. I pjust wanted to show all the options before removing them.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java (hunk 4 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java]: @gsmet told me it is an ES thing. It did surprise me too but they do expose unit for some reason.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java (hunk 5 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: See comment above.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java (hunk 6 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: Today when you do a spatial query and want to sort by it, you need to duplicate both the field name and the coordinates to distance from. This API goal is to avoid the duplication concern. Looks at the JIRA, I did show a more complete exampl in context.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java (hunk 7 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: I,ve seen it in ES examples, I'm not sure it is exotic.

Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: Score and field I could be convinced of the exotism. Field then score, less so :)

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java (hunk 9 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java]: Right I was thinking of offering alternative implementations.

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java (hunk 10 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java]: Hum, elastic search does not document their default if they even have one. I imagine they pick the first unless specified ?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java (hunk 2 above)
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: I guess the question is if you expect the existing "setSort" to be removed in future, and be replaced by this DSL as exclusive way?

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java (hunk 12 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java]: I actually copy pasted the ES doc here. These are their limitations. Maybe we should not expose this in the first round.

File: orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java (hunk 13 above)
Comment (emmanuelbernard) [inline orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java]: This is not my code, I copied it :)

File: orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java (hunk 14 above)
Comment (emmanuelbernard) [inline orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java]: I copied the DSLTest code really

File: orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java (hunk 13 above)
Comment (Sanne) [inline orm/src/test/java/org/hibernate/search/test/query/dsl/SortDSLTest.java]: I knew it! your code is usually better :)

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java (hunk 2 above)
Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: if we manage to eliminate `createFullTextQuery` when a query is built from the DSL, then we should do the same for `setSort`.

Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: Ok that sounds compelling, but currently we can't eliminate `createFullTextQuery` as our query DSL only covers building a handful of queries.

So your suggestion is that in a next step we'd need to provide a full alternative to `createFullTextQuery` able to produce any advanced & custom query... essentially a way to include "native query definitions" in our DSL in a similar and consistent way with the Sort DSL proposal?

That sounds nice and I'm not against it, but I'm surprised that you'd start such a direction from the sort DSL.

Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: The DSL is already native friendly. A boolean query for example can aggregate native Lucene queries. This is very much like the native integration of native `SortField`s in sort. Where you thinking of something else?

Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: Ok fair enough. Just trying to gauge were this is going, as I expected a much simpler Sort API.. but I guess you're right I hadn't though about this aspect.

Let's try to make concrete preparations for deprecating the traditional `createFullTextQuery` during the 6.0 works then.

Comment (emmanuelbernard) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortNativeContext.java]: Just to be clear, I don't have a clue how the replacement of `createFullTextQuery` would look like :D

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java (hunk 4 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortDistanceContext.java]: @Sanne when you do a sort by distance, ES includes the distance calculated in the results so that you don't have to calculate it yourself. This is why they expose a way to tweak the unit. We use it in our ES work to avoid using a script to calculate the distance.

See https://github.com/hibernate/hibernate-search/blob/master/elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchHSQueryImpl.java#L541 .

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java (hunk 7 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/SortAdditionalSortFieldContext.java]: I don't think score and field are so exotic. You might use a scoring which leads to equal scores and want to get stable results (and add an order by name for instance).

File: engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java (hunk 12 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/query/dsl/sort/MultiValuedMode.java]: Yup. Looks like it's orthogonal to the current work. We will need to dig in ES to see how they do it and port it to our Lucene backend. Better let it aside for now.

**Comment by gsmet** [comment]
Closing for now. Yoann is working on a new PR superseding this one.
